File: Data Structures/BST Insert and Remove/Implementation.go
package main

// TreeNode represents a node in a binary search tree.
type TreeNode struct {
    Val   int       // Value of the node
    Left  *TreeNode // Pointer to the left child
    Right *TreeNode // Pointer to the right child
}

// NewTreeNode creates and returns a new TreeNode with the given value.
func NewTreeNode(val int) *TreeNode {
    return &TreeNode{
        Val: val,
    }
}

// InsertAndRemove struct doesn't hold any data but provides methods for BST operations.
type InsertAndRemove struct {
}

// Insert a new node with the given value into the BST and return the root of the modified tree.
func (ir *InsertAndRemove) Insert(root *TreeNode, val int) *TreeNode {
    // If the current node is nil, create a new node with the value.
    if root == nil {
        return NewTreeNode(val)
    }

    // If the value is greater than the current node's value, insert it into the right subtree.
    if val > root.Val {
        root.Right = ir.Insert(root.Right, val)
    } else if val < root.Val { // If the value is less, insert it into the left subtree.
        root.Left = ir.Insert(root.Left, val)
    }
    // Return the root of the tree after insertion.
    return root
}

// MinValueNode returns the node with the minimum value found in the BST.
func (ir *InsertAndRemove) MinValueNode(root *TreeNode) *TreeNode {
    curr := root
    // Traverse to the leftmost node, which has the minimum value.
    while curr != nil && curr.Left != nil {
        curr = curr.Left
    }
    return curr
}

// Remove a node with the given value from the BST and return the root of the modified tree.
func (ir *InsertAndRemove) Remove(root *TreeNode, val int) *TreeNode {
    // If the root is nil, the tree is empty, return nil.
    if root == nil {
        return nil
    }

    // If the value to be removed is greater than the root's value, search in the right subtree.
    if val > root.Val {
        root.Right = ir.Remove(root.Right, val)
    } else if val < root.Val { // If less, search in the left subtree.
        root.Left = ir.Remove(root.Left, val)
    } else {
        // Node with only one child or no child
        if root.Left == nil {
            return root.Right
        } else if root.Right == nil {
            return root.Left
        }

        // Node with two children: Get the inorder successor (smallest in the right subtree)
        minNode := ir.MinValueNode(root.Right)

        // Copy the inorder successor's content to this node
        root.Val = minNode.Val

        // Delete the inorder successor
        root.Right = ir.Remove(root.Right, minNode.Val)
    }
    // Return the root of the tree after deletion.
    return root
}

// time complexity: O(h), where h is the height of the tree.
// space complexity: O(h), where h is the height of the tree.

File: Data Structures/Trees/BST Insert and Remove/Implementation.py
class TreeNode:
    # Constructor to initialize a node
    def __init__(self, val):
        self.val = val  # Value of the node
        self.left = None  # Pointer to the left child
        self.right = None  # Pointer to the right child

# Function to insert a new node with a given value into the BST.
def insert(root, val):
    # If the current node is None, create a new node with the value and return it.
    if not root:
        return TreeNode(val)
    
    # If the value to insert is greater than the current node's value, insert it into the right subtree.
    if val > root.val:
        root.right = insert(root.right, val)
    # If the value to insert is less than the current node's value, insert it into the left subtree.
    elif val < root.val:
        root.left = insert(root.left, val)
    # Return the (possibly updated) current node.
    return root

# Function to find the node with the minimum value in the BST.
def minValueNode(root):
    curr = root  # Start with the current node as the root.
    # Loop to find the leftmost leaf, which is the minimum value node in the BST.
    while curr and curr.left:
        curr = curr.left
    # Return the node with the minimum value.
    return curr

# Function to remove a node with a given value from the BST.
def remove(root, val):
    # If the root is None, the tree is empty, so return None.
    if not root:
        return None
    
    # If the value to remove is greater than the current node's value, go to the right subtree.
    if val > root.val:
        root.right = remove(root.right, val)
    # If the value to remove is less than the current node's value, go to the left subtree.
    elif val < root.val:
        root.left = remove(root.left, val)
    else:
        # If the node to be deleted is a leaf or has one child, replace it with its child.
        if not root.left:
            return root.right
        elif not root.right:
            return root.left
        # If the node has two children, replace it with its in-order successor.
        else:
            minNode = minValueNode(root.right)  # Find the in-order successor.
            root.val = minNode.val  # Copy the in-order successor's value to this node.
            root.right = remove(root.right, minNode.val)  # Delete the in-order successor.
    # Return the (possibly updated) root.
    return root

# time complexity: O(h), where h is the height of the BST.
# space complexity: O(h), where h is the height of the BST.